File: 프로그래머스/2/131127. 할인 행사/할인 행사.cpp
#include <string>
#include <vector>
#include <unordered_map>
#include <iostream>

using namespace std;

int solution(vector<string> want, vector<int> number, vector<string> discount) {
    int answer = 0;
    
    unordered_map<string, int> um1;
    
    for(int i = 0; i < want.size(); i++){
        um1[want[i]] = number[i];
    }
    
    for(int i = 0; i < discount.size(); i++){
        if(i + 10 <= discount.size()){
            unordered_map<string, int> um2;
            bool flag = true;
            
            for(int j = i; j < i + 10; j++){
                um2[discount[j]]++;
            }
            
            for(const auto &item: um1){
                if(um2[item.first] != item.second){
                    flag = false;
                    break;
                }
            }
            
            if(flag) answer++;
        }
    }
    
    return answer;
}

File: 프로그래머스/2/131127. 할인 행사/할인 행사.py
def solution(want, number, discount):
    answer = 0
    wants = {want[i]: number[i] for i in range(len(want))}
    
    for i in range(len(discount) - 9):
        wants_copied = wants.copy()
        
        for j in range(i, i + 10):
            if discount[j] in wants_copied:
                wants_copied[discount[j]] -= 1
        
        if all(count == 0 for count in wants_copied.values()): # sum(wants_copied.values() == 0의 반례, [1, -1, 0, 0, 0]
            answer += 1
        
    return answer